Mathlib/RingTheory/TwoSidedIdeal/Basic.lean

```diff
@@ -0,0 +1,172 @@
+
+section NonUnitalNonAssocRing
+
+variable {R : Type*} [NonUnitalNonAssocRing R] (I : TwoSidedIdeal R)
+
+instance setLike : SetLike (TwoSidedIdeal R) R where
+  coe t := {r | t.ringCon r 0}
+  coe_injective'  := by
+    rintro ⟨t₁⟩ ⟨t₂⟩ (h : {x | _} = {x | _})
+    congr 1
+    refine RingCon.ext fun a b ↦ ⟨fun H ↦ ?_, fun H ↦ ?_⟩
+    · have H' : a - b ∈ {x | t₁ x 0} := sub_self b ▸ t₁.sub H (t₁.refl b)
+      rw [h] at H'
+      convert t₂.add H' (t₂.refl b) using 1 <;> abel
+    · have H' : a - b ∈ {x | t₂ x 0} := sub_self b ▸ t₂.sub H (t₂.refl b)
+      rw [← h] at H'
+      convert t₁.add H' (t₁.refl b) using 1 <;> abel
+
+lemma mem_iff (x : R) : x ∈ I ↔ I.ringCon x 0 := Iff.rfl
+
+lemma rel_iff (x y : R) : I.ringCon x y ↔ x - y ∈ I := by
+  rw [mem_iff]
+  constructor
+  · intro h; convert I.ringCon.sub h (I.ringCon.refl y); abel
+  · intro h; convert I.ringCon.add h (I.ringCon.refl y) <;> abel
+
+/--
+the coercion from two-sided-ideals to sets is an order embedding
+-/
+@[simps]
+def coeOrderEmbedding : TwoSidedIdeal R ↪o Set R where
+  toFun := SetLike.coe
+  inj' := SetLike.coe_injective
+  map_rel_iff' {I J} := ⟨fun (h : (I : Set R) ⊆ (J : Set R)) _ h' => h h', fun h _ h' => h h'⟩
+
+lemma le_iff {I J : TwoSidedIdeal R} : I ≤ J ↔ (I : Set R) ⊆ (J : Set R) :=
+  coeOrderEmbedding.map_rel_iff.symm
+
+@[ext]
+lemma ext {I J : TwoSidedIdeal R} (h : ∀ x, x ∈ I ↔ x ∈ J) : I = J :=
+  coeOrderEmbedding.injective (Set.ext h)
+
+lemma lt_iff (I J : TwoSidedIdeal R) : I < J ↔ (I : Set R) ⊂ (J : Set R) := by
+  rw [lt_iff_le_and_ne, Set.ssubset_iff_subset_ne, le_iff]
+  simp
+
+lemma zero_mem : 0 ∈ I := I.ringCon.refl 0
+
+lemma add_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x + y ∈ I := by simpa using I.ringCon.add hx hy
+
+lemma neg_mem {x} (hx : x ∈ I) : -x ∈ I := by simpa using I.ringCon.neg hx
+
+lemma sub_mem {x y} (hx : x ∈ I) (hy : y ∈ I) : x - y ∈ I := by simpa using I.ringCon.sub hx hy
+
+lemma mul_mem_left (x y) (hy : y ∈ I) : x * y ∈ I := by
+  simpa using I.ringCon.mul (I.ringCon.refl x) hy
+
+lemma mul_mem_right (x y) (hx : x ∈ I) : x * y ∈ I := by
+  simpa using I.ringCon.mul hx (I.ringCon.refl y)
+
+lemma nsmul_mem {x} (n : ℕ) (hx : x ∈ I) : n • x ∈ I := by simpa using I.ringCon.nsmul _ hx
+
+lemma zsmul_mem {x} (n : ℤ) (hx : x ∈ I) : n • x ∈ I := by simpa using I.ringCon.zsmul _ hx
+
+/--
+The "set-theoretic-way" of constructing a two-sided ideal by providing:
+- the underlying set `S`;
+- a proof that `0 ∈ S`;
+- a proof that `x + y ∈ S` if `x ∈ S` and `y ∈ S`;
+- a proof that `-x ∈ S` if `x ∈ S`;
+- a proof that `x * y ∈ S` if `y ∈ S`;
+- a proof that `x * y ∈ S` if `x ∈ S`.
+-/
+def mk' (carrier : Set R)
+    (zero_mem : 0 ∈ carrier)
+    (add_mem : ∀ {x y}, x ∈ carrier → y ∈ carrier → x + y ∈ carrier)
+    (neg_mem : ∀ {x}, x ∈ carrier → -x ∈ carrier)
+    (mul_mem_left : ∀ {x y}, y ∈ carrier → x * y ∈ carrier)
+    (mul_mem_right : ∀ {x y}, x ∈ carrier → x * y ∈ carrier) : TwoSidedIdeal R where
+  ringCon :=
```

Comment:
Since we are using a structure, might as well provide an alternative constructor